### PR TITLE
Better asm syntax highlighting

### PR DESCRIPTION
**Pull request by lievenhey**

### CODE REVIEW

**Comment by lievenhey** [comment]
left old, right new
grafik

**Comment by lievenhey** [comment]
TODO:
- test
- hide format chooser if ansi codes are used
- add option to hide hexdump

**Comment by GitMensch** [comment]
I'm a bit confused, why does "Syntax Highlighting: None" have any color? But nice to see that those issues are tackled!

**Comment by lievenhey** [comment]
> I'm a bit confused, why does "Syntax Highlighting: None" have any color? But nice to see that those issues are tackled!

someone didn't implement synchronisation yet

**Comment by GitMensch** [comment]
As this auto-drops the highlighting on the hex values and also changes the code involved here, I highly suggest to implement #527 (moving the hex parts out as requested by Milian) first. Could be done in this PR, too.

File: src/models/disassemblyoutput.cpp [hunk 1]
```diff
@@ -207,29 +207,25 @@ DisassemblyOutput::ObjectdumpOutput DisassemblyOutput::objdumpParse(const QByteA
...
-            const auto addrEnd = lineRest.indexOf(suffix, prefix.size());
-            if (addrEnd == -1)
+        const auto addr = [addrString = parts[0].trimmed(), &asmLine]() -> uint64_t {
+            const auto suffix = QLatin1Char(':');
+            if (!addrString.endsWith(suffix))
                 return 0;
 
             bool ok = false;
-            auto ret = lineRest.mid(0, addrEnd).toULongLong(&ok, 16);
+            auto ret = addrString.midRef(0, addrString.size() - 1).toULongLong(&ok, 16);
```
Comment (milianw) [inline src/models/disassemblyoutput.cpp:249]: the midRef isn't needed anymore now, it should be the same as just `addrString.toULongLong(...)`

File: src/models/disassemblyoutput.cpp [hunk 2]
```diff
@@ -207,29 +207,25 @@ DisassemblyOutput::ObjectdumpOutput DisassemblyOutput::objdumpParse(const QByteA
...
-                return 0;
+        // a line looks like this:
+        // [spaces]addr:\t [branch visualisazion] [hexdump]\tdiassembly
+        // we can simplify parsing by splitting it into three parts
+
+        const auto parts = asmLine.split(QLatin1Char('\t'));
 
-            const auto addrEnd = lineRest.indexOf(suffix, prefix.size());
-            if (addrEnd == -1)
+        const auto addr = [addrString = parts[0].trimmed(), &asmLine]() -> uint64_t {
```
Comment (milianw) [inline src/models/disassemblyoutput.cpp:243]: `parts.value(0)` for extra safety

File: src/models/disassemblyoutput.cpp [hunk 3]
```diff
@@ -238,19 +234,23 @@ DisassemblyOutput::ObjectdumpOutput DisassemblyOutput::objdumpParse(const QByteA
...
-            auto size = std::distance(lineRest.cbegin(), firstHexIt);
-            auto ret = lineRest.mid(0, size);
-            lineRest = lineRest.mid(size);
-            return ret;
+        struct BranchesAndHexdump
+        {
+            QString branchVisualisation;
+            QString hexdump;
+        };
+        const auto [branchVisualisation, hexdump] = [branchesAndHex = parts[1]]() -> BranchesAndHexdump {
```
Comment (milianw) [inline src/models/disassemblyoutput.cpp:269]: `parts.value(1)` for extra safety

File: src/models/disassemblyoutput.cpp [hunk 4]
```diff
@@ -238,19 +234,23 @@ DisassemblyOutput::ObjectdumpOutput DisassemblyOutput::objdumpParse(const QByteA
...
+            auto size = std::distance(branchesAndHex.cbegin(), firstHexIt);
+            auto branchVisualisation = branchesAndHex.mid(0, size);
+            auto hexdump = branchesAndHex.mid(size);
+            return {branchVisualisation, hexdump};
         }();
 
         disassemblyLines.push_back({addr,
-                                    lineRest.toString(),
-                                    branchVisualisation.toString(),
+                                    parts.size() == 3 ? parts[2].trimmed() : QString {},
```
Comment (milianw) [inline src/models/disassemblyoutput.cpp:278]: simplify to `parts.value(2).trimmed()`

File: src/models/disassemblymodel.cpp [hunk 5]
```diff
@@ -82,6 +82,8 @@ QVariant DisassemblyModel::headerData(int section, Qt::Orientation orientation,
         return tr("Address");
     else if (section == BranchColumn)
         return tr("Branches");
+    else if (section == HexdumpColumn)
```
Comment (milianw) [inline src/models/disassemblymodel.cpp:74]: this patch is missing test coverage

File: src/models/disassemblyoutput.cpp [hunk 6]
```diff
@@ -17,28 +17,46 @@
 #include <QProcess>
 #include <QStandardPaths>
 
+#include "formattingutils.h"
+
 namespace {
 Q_LOGGING_CATEGORY(disassemblyoutput, "hotspot.disassemblyoutput")
 
-bool canVisualizeJumps(const QString& objdump)
+QString objdumpHelp(const QString& objdump)
```
Comment (milianw) [inline src/models/disassemblyoutput.cpp:25]: just return a `QByteArray` and use that then with `QByteArrayLiteral` - then we don't need the extra `QString` conversion

File: src/models/disassemblyoutput.cpp [hunk 7]
```diff
@@ -17,28 +17,46 @@
...
+{
+    return objdumpHelp(objdump).contains(QStringLiteral("--disassembler-color"));
 }
 
-DisassemblyOutput::LinkedFunction extractLinkedFunction(const QString& disassembly)
+DisassemblyOutput::LinkedFunction extractLinkedFunction(const QString& disassemblyWithAnsi)
 {
     DisassemblyOutput::LinkedFunction function = {};
 
+    const auto escapeChar = QLatin1Char('\u001B');
```
Comment (milianw) [inline src/models/disassemblyoutput.cpp:52]: add a comment - what is this char?

File: src/models/disassemblyoutput.cpp [hunk 8]
```diff
@@ -213,6 +232,14 @@ DisassemblyOutput::ObjectdumpOutput DisassemblyOutput::objdumpParse(const QByteA
 
         const auto parts = asmLine.split(QLatin1Char('\t'));
 
+        if (parts.size() == 1 && asmLine.endsWith(QLatin1Char(':'))) {
```
Comment (milianw) [inline src/models/disassemblyoutput.cpp:235]: it's unfortunate that you do so much in this patch - I would really appreciate if you could split it up - one patch that just ports away from the old QTextDocument usage to the new highlighter, then another patch for adding the custom highlighting

File: src/models/formattingutils.h [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+    SPDX-FileCopyrightText: 2023 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+
+    SPDX-License-Identifier: GPL-2.0-or-later
+*/
+
+#pragma once
+
+#include <QString>
+
+namespace utils {
```
Comment (milianw) [inline src/models/formattingutils.h:12]: afair we use `Util` in all other areas already, use the same casing here too

File: src/models/formattingutils.h [hunk 10]
```diff
@@ -0,0 +1,14 @@
+/*
+    SPDX-FileCopyrightText: Lieven Hey <lieven.hey@kdab.com>
+    SPDX-FileCopyrightText: 2023 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+
+    SPDX-License-Identifier: GPL-2.0-or-later
+*/
+
+#pragma once
+
+#include <QString>
```
Comment (milianw) [inline src/models/formattingutils.h:10]: forward declare

File: src/models/highlightedtext.cpp [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+    {
+        m_formats.clear();
+        m_offset = 0;
+
+        KSyntaxHighlighting::State state;
+        for (const auto& line : text) {
+            state = highlightLine(line, state);
+
+            // KSyntaxHighlighting uses line offsets but QTextLayout uses global offsets
+            // the +1 is for the newline at the end of the line
```
Comment (milianw) [inline src/models/highlightedtext.cpp:49]: there is no +1 here

File: src/models/highlightedtext.cpp [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+    m_layout->setCacheEnabled(true);
+    m_layout->setFont(QFontDatabase::systemFont(QFontDatabase::FixedFont));
+}
+
+HighlightedText::~HighlightedText() = default;
+
+void HighlightedText::setText(const QStringList& text)
+{
+    const bool usesAnsi = std::any_of(text.cbegin(), text.cend(),
+                                      [](const QString& line) { return line.contains(QLatin1Char('\u001B')); });
```
Comment (milianw) [inline src/models/highlightedtext.cpp:198]: again this magic char, what does it mean? if it has to be in sync with the other place, introduce it as a constexpr global with a sane name

File: src/resultsdisassemblypage.cpp [hunk 13]
```diff
@@ -177,9 +177,11 @@ class BranchDelegate : public QStyledItemDelegate
 
         auto pen = QPen(option.palette.color(QPalette::Link), 1);
         pen.setCosmetic(true);
+        const auto oldPen = painter->pen();
```
Comment (milianw) [inline src/resultsdisassemblypage.cpp:180]: the commit message for this patch has an error - duplicate `will will`

**Review by milianw — CHANGES_REQUESTED** [review]
the cleanup/fixup patches should go into a separate MR because they can land directly. the rest of this patch needs some cleanup

File: src/models/disassemblymodel.cpp (hunk 5 above)
Comment (lievenhey) [reply src/models/disassemblymodel.cpp:74]: ah it moved into the next patch, dammit

File: src/models/highlightedtext.cpp (hunk 12 above)
Comment (lievenhey) [reply src/models/highlightedtext.cpp:198]: well its know as escape character
if you look at the ascii table you can see it there
```
       032   26    1A    SUB (substitute)            132   90    5A    Z
       033   27    1B    ESC (escape)                133   91    5B    [
       034   28    1C    FS  (file separator)        134   92    5C    \  '\\'
```

Comment (milianw) [reply src/models/highlightedtext.cpp:195]: you cannot expect people have the ASCII table in their mind when reading the code. the point is: your code is unreadable as-is. make it readable by introducing a named variable:

```
constexpr auto escapeCode = QLatin1Char('\u001B');
```

File: tests/modeltests/tst_disassemblyoutput.cpp [hunk 14]
```diff
@@ -207,30 +207,32 @@ private slots:
...
+        };
+
         for (const auto& line : result.disassemblyLines) {
             QVERIFY(!line.branchVisualisation.isEmpty());
 
             // check that we only captures valid visualisation characters
             QVERIFY(std::all_of(line.branchVisualisation.cbegin(), line.branchVisualisation.cend(),
                                 isValidVisualisationCharacter));
 
-            QVERIFY(!line.disassembly.isEmpty());
```
Comment (milianw) [inline tests/modeltests/tst_disassemblyoutput.cpp:225]: why did you remove the old code that checks that the disassembly wasn't empty? that should still be the case, no?

File: src/models/disassemblyoutput.cpp [hunk 15]
```diff
@@ -17,28 +17,43 @@
 #include <QProcess>
 #include <QStandardPaths>
 
+#include "formattingutils.h"
+
 namespace {
 Q_LOGGING_CATEGORY(disassemblyoutput, "hotspot.disassemblyoutput")
 
-bool canVisualizeJumps(const QString& objdump)
+QByteArray objdumpHelp(const QString& objdump)
```
Comment (milianw) [inline src/models/disassemblyoutput.cpp:25]: unrelated cleanup -> separate patch

File: src/models/disassemblyoutput.cpp [hunk 16]
```diff
@@ -17,28 +17,43 @@
...
-    return help.contains("--visualize-jumps");
+    return process.readAllStandardOutput();
+}
+
+bool canVisualizeJumps(const QString& objdump)
+{
+    return objdumpHelp(objdump).contains(QByteArrayLiteral("--visualize-jumps"));
+}
+
+bool canUseSyntaxHighlighting(const QString& objdump)
```
Comment (milianw) [inline src/models/disassemblyoutput.cpp:42]: the patch that introduces disassembler highlighting should be a separate one again

File: src/models/disassemblyoutput.cpp (hunk 8 above)
Comment (milianw) [reply src/models/disassemblyoutput.cpp:232]: this wasn't done yet?

File: src/models/highlightedtext.cpp [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+    }
+
+    void themeChanged() override
+    {
+        if (!m_repository) {
+            return;
+        }
+
+        KSyntaxHighlighting::Repository::DefaultTheme theme;
+        if (QPalette().base().color().lightness() < 128) {
```
Comment (milianw) [inline src/models/highlightedtext.cpp:62]: this cannot work, can it? see: https://doc.qt.io/qt-6/qpalette.html#QPalette

you want to use https://doc.qt.io/qt-6/qguiapplication.html#palette or something instead, no?

File: tests/modeltests/tst_formatting.cpp [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+class TestFormatting : public QObject
+{
+    Q_OBJECT
+private slots:
+    void testRemoveAnsi_data()
+    {
+        QTest::addColumn<QString>("ansiString");
+        QTest::addColumn<QString>("ansiFreeString");
+
+        QTest::addRow("no ansi sequence") << QStringLiteral("[30m A B [0m C") << QStringLiteral("[30m A B [0m C");
```
Comment (milianw) [inline tests/modeltests/tst_formatting.cpp:25]: ansi highlighting should be done in a separate patch

File: src/models/highlightedtext.cpp (hunk 12 above)
Comment (milianw) [reply src/models/highlightedtext.cpp:195]: you even introduced this now (in the wrong patch) - use it here too

**Review by milianw — CHANGES_REQUESTED** [review]

File: src/models/highlightedtext.cpp (hunk 17 above)
Comment (lievenhey) [reply src/models/highlightedtext.cpp:62]: good to know, I copied that code from another location, will fix both

File: tests/modeltests/tst_disassemblyoutput.cpp (hunk 14 above)
Comment (lievenhey) [reply tests/modeltests/tst_disassemblyoutput.cpp:225]: no, sometimes we encounter `nop`s like this:
```asm
    124a:       f2 0f 10 4c 24 08       movsd  0x8(%rsp),%xmm1
    1250:       f2 0f 10 44 24 10       movsd  0x10(%rsp),%xmm0
    1256:       f2 0f 58 0d aa 0d 00    addsd  0xdaa(%rip),%xmm1        # 2008 <_IO_stdin_used+0x8>
    125d:       00 
    125e:       ff c3                   inc    %ebx
    1260:       81 fb 40 42 0f 00       cmp    $0xf4240,%ebx
    1266:       0f 85 64 ff ff ff       jne    11d0 <main+0x50>
```

Comment (GitMensch) [reply tests/modeltests/tst_disassemblyoutput.cpp:257]: That's important information, maybe add here as comment?

**Comment by GitMensch** [comment]
This also fixes #527 (currently not related, you may want to do that on the commit message).

Can you please include the logs https://github.com/KDAB/hotspot/files/13114752/dump.log and https://github.com/KDAB/hotspot/files/13114535/log.txt in the testsuite to ensure they can be processed?

File: tests/modeltests/tst_disassemblyoutput.cpp (hunk 14 above)
Comment (milianw) [reply tests/modeltests/tst_disassemblyoutput.cpp:257]: I agree, add a comment - better yet explicitly catch it - i.e. expect empty when the hexdump is `00`, otherwise expect non-empty

File: src/models/highlightedtext.cpp [hunk 19]
```diff
@@ -0,0 +1,202 @@
...
+#include <KSyntaxHighlighting/State>
+#include <KSyntaxHighlighting/SyntaxHighlighter>
+#include <KSyntaxHighlighting/Theme>
+#endif
+
+#include "formattingutils.h"
+
+namespace {
+#if KFSyntaxHighlighting_FOUND
+class KSyntaxHighlightingImplementation : public HighlightingImplementation,
```
Comment (milianw) [inline src/models/highlightedtext.cpp:30]: don't use the `K` prefix here, to ensure we don't run into name clashes in the future

File: src/models/highlightedtext.h [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+class Definition;
+class Repository;
+}
+
+class HighlightingImplementation
+{
+    Q_DISABLE_COPY(HighlightingImplementation)
+public:
+    HighlightingImplementation() = default;
+    virtual ~HighlightingImplementation() = default;
```
Comment (milianw) [inline src/models/highlightedtext.h:29]: move the `= default` to the .cpp - we don't want to inline the vtable into every TU that includes this header

File: src/models/highlightedtext.h [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+    SPDX-License-Identifier: GPL-2.0-or-later
+*/
+
+#pragma once
+
+#include <QObject>
+#include <QTextLayout>
+#include <QTextLine>
+
+#include <KColorScheme>
```
Comment (milianw) [inline src/models/highlightedtext.h:14]: should be unused in the header?

File: src/models/highlightedtext.h [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+    SPDX-FileCopyrightText: Lieven Hey <lieven.hey@kdab.com>
+    SPDX-FileCopyrightText: 2023 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+
+    SPDX-License-Identifier: GPL-2.0-or-later
+*/
+
+#pragma once
+
+#include <QObject>
+#include <QTextLayout>
```
Comment (milianw) [inline src/models/highlightedtext.h:11]: forward declare

File: src/models/highlightedtext.h [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+#include <QTextLine>
```
Comment (milianw) [inline src/models/highlightedtext.h:12]: dito

File: src/models/highlightedtext.h [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+private slots:
+    void applyFormatting();
+
+private:
+    void updateHighlighting();
+
+#if KFSyntaxHighlighting_FOUND
+    KSyntaxHighlighting::Repository* m_repository;
+#endif
+    std::unique_ptr<HighlightingImplementation> m_highlighter;
```
Comment (milianw) [inline src/models/highlightedtext.h:87]: missing <memory> include?

File: src/models/highlightedtext.cpp [hunk 25]
```diff
@@ -0,0 +1,202 @@
...
+
+    SPDX-License-Identifier: GPL-2.0-or-later
+*/
+
+#include "highlightedtext.h"
+
+#include <QGuiApplication>
+#include <QPalette>
+
+#include <memory>
```
Comment (milianw) [inline src/models/highlightedtext.cpp:13]: move to header

File: src/models/highlightedtext.cpp [hunk 26]
```diff
@@ -0,0 +1,202 @@
...
+
+#include <KColorScheme>
```
Comment (milianw) [inline src/models/highlightedtext.cpp:15]: empty line below

File: src/models/highlightedtext.h [hunk 27]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+class HighlightingImplementation
+{
+    Q_DISABLE_COPY(HighlightingImplementation)
+public:
+    HighlightingImplementation() = default;
+    virtual ~HighlightingImplementation() = default;
+
+    virtual QVector<QTextLayout::FormatRange> format(const QStringList& text) = 0;
```
Comment (milianw) [inline src/models/highlightedtext.h:31]: why is this a pure virtual interface? there is only one implementation at compile time - remove all the virtuals here from the header

Comment (milianw) [reply src/models/highlightedtext.h:31]: also, this class is not public - meaning people should only deal with HighlightedText below which uses it internally but otherwise doesn't provide a getter to this class. As such, forward declare this here in the header and move all code to the .cpp

File: src/models/highlightedtext.cpp (hunk 19 above)
Comment (milianw) [reply src/models/highlightedtext.cpp:30]: see comments in the header: I don't understand why you have this indirection at all. Just make this the `class HighlightingImplementation` and be done with it - no need for multiple inheritance or the `KSyntaxHighlightingImplementation` at all

File: src/models/disassemblyoutput.cpp [hunk 28]
```diff
@@ -20,17 +20,21 @@
...
+        return {};
     }
-    const auto help = process.readAllStandardOutput();
-    return help.contains("--visualize-jumps");
+    return process.readAllStandardOutput();
+}
+
+bool canVisualizeJumps(const QString& objdump)
+{
+    return objdumpHelp(objdump).contains(QByteArrayLiteral("--visualize-jumps"));
```
Comment (milianw) [inline src/models/disassemblyoutput.cpp:39]: sorry, I was wrong - the QBAL isn't needed here (see the old code - just `.contains("--visualize-jumps")`

File: src/models/highlightedtext.cpp [hunk 29]
```diff
@@ -0,0 +1,202 @@
...
+#include <KSyntaxHighlighting/AbstractHighlighter>
+#include <KSyntaxHighlighting/Definition>
+#include <KSyntaxHighlighting/Format>
+#include <KSyntaxHighlighting/Repository>
+#include <KSyntaxHighlighting/State>
+#include <KSyntaxHighlighting/SyntaxHighlighter>
+#include <KSyntaxHighlighting/Theme>
+#endif
+
+#include "formattingutils.h"
```
Comment (milianw) [inline src/models/highlightedtext.cpp:26]: the .cpp for this is only added to CMakeLists.txt in a later patch

File: src/models/CMakeLists.txt [hunk 30]
```diff
@@ -11,6 +11,7 @@ add_library(
     disassemblyoutput.cpp
     eventmodel.cpp
     filterandzoomstack.cpp
+    formattingutils.cpp
```
Comment (milianw) [inline src/models/CMakeLists.txt:14]: a previous patch already used the header for this

File: src/models/disassemblyoutput.cpp [hunk 31]
```diff
@@ -37,12 +39,21 @@ bool canVisualizeJumps(const QString& objdump)
     return objdumpHelp(objdump).contains(QByteArrayLiteral("--visualize-jumps"));
 }
 
-DisassemblyOutput::LinkedFunction extractLinkedFunction(const QString& disassembly)
+bool canUseSyntaxHighlighting(const QString& objdump)
+{
+    return objdumpHelp(objdump).contains(QByteArrayLiteral("--disassembler-color"));
```
Comment (milianw) [inline src/models/disassemblyoutput.cpp:44]: sorry, dito - no QBAL needed

File: src/models/highlightedtext.cpp (hunk 12 above)
Comment (milianw) [reply src/models/highlightedtext.cpp:197]: this isn't done

File: tests/modeltests/tst_disassemblyoutput.cpp [hunk 32]
```diff
@@ -126,6 +126,7 @@ private slots:
         }
 
         file->write(text.toUtf8());
+        file->flush();
```
Comment (milianw) [inline tests/modeltests/tst_disassemblyoutput.cpp:129]: why was this not detected on the CI so far?

**Review by milianw — CHANGES_REQUESTED** [review]

File: tests/modeltests/tst_disassemblyoutput.cpp (hunk 32 above)
Comment (lievenhey) [reply tests/modeltests/tst_disassemblyoutput.cpp:129]: :shrug: it worked on older versions, so I guess ether the host or the ci container got an update which changed  how the write calls are handled

**Comment by GitMensch** [comment]
Rechecked: while the appimage 211 showed the syntax-highlighted disassembly and used "non-7bit arrows" (and a warning message `unexepected jump visualization char: +`)
highlight
the current appimage has seems to have no syntax-highlighting any more
no-highlight
along with the warning
```
hotspot.disassemblyoutput: objdump binary does not support `--disassembler-color`: "/usr/bin/objdump"
```
changing the highlighter in the pull-down doesn't seem to have any effect at all

File: src/models/disassemblyoutput.cpp [hunk 33]
```diff
@@ -238,19 +261,23 @@ DisassemblyOutput::ObjectdumpOutput DisassemblyOutput::objdumpParse(const QByteA
...
+        {
+            QString branchVisualisation;
+            QString hexdump;
+        };
+        const auto [branchVisualisation, hexdump] = [branchesAndHex = parts.value(1)]() -> BranchesAndHexdump {
+            auto firstHexIt = std::find_if(branchesAndHex.cbegin(), branchesAndHex.cend(), isHexCharacter);
+            auto size = std::distance(branchesAndHex.cbegin(), firstHexIt);
+            auto branchVisualisation = branchesAndHex.mid(0, size);
+            auto hexdump = branchesAndHex.mid(size);
+            return {branchVisualisation, hexdump};
```
Comment (GitMensch) [inline src/models/disassemblyoutput.cpp:274]: Can you use `return {branchVisualisation, hexdump..trimmed()};` here? Otherwise we have trailing spaces, leading to eliding symbol "..." (and unnecessary amount of trailing spaces kept in memory).

**Review by GitMensch — CHANGES_REQUESTED** [review]

**Comment by lievenhey** [comment]
> ```
> hotspot.disassemblyoutput: objdump binary does not support `--disassembler-color`: "/usr/bin/objdump"
> ```
that error should definitely not happen, since I check if that flag is supported, I will take another look at the code

**Comment by GitMensch** [comment]
Seems like the cmake-format CI error is easy to solve, but the failing test cases need more work.

**Comment by GitMensch** [comment]
Just wanted to share that #561 seems to be fixed with the last commit (only checked two _different_ examples that did not work before) - if it should be then please add its reference to the commits / this PR.
Can you add the disassembly output provided in this issue to the testsuite?

One downside: it seems that the syntax highlighting in the Disassembly does not work (AppImage).

**Comment by lievenhey** [comment]
@GitMensch can check you the most recent appimage? I notices that QTextLayout is very slow (multiple seconds for 50k+ lines) and added a cache. You complained about slow performance so maybe this finally fixes it.

**Comment by GitMensch** [comment]
Checked one extreme example (all with appimages):

* "old" hotspot: ~110 seconds
* current master: 48-60 seconds
* this PR previous push: more seconds
* this PR after your last commit: 11-12 seconds - but the syntax highlighting in the disassembly is "kind of gone", compare
   master with
   this PR's state

**Comment by lievenhey** [comment]
That is most likely, because `KColorScheme::ColorSet::Complementary` is not available. I will think of another way to get colors.

**Comment by GitMensch** [comment]
> That is most likely, because `KColorScheme::ColorSet::Complementary` is not available. I will think of another way to get colors.

... and/or because of the following?

> Note that I do get a message
> 
> > hotspot.disassemblyoutput: objdump binary does not support `--disassembler-color`: "/usr/bin/objdump"

**Comment by lievenhey** [comment]
Ohh, right, in that case it should fallback to using `KSyntaxHighlighter`. In that case it should use the original formatting. I will into it.

**Comment by lievenhey** [comment]
@GitMensch I looked into it. Currently I only copy the font color from `KSyntaxHighlighter::Format`. If that is what you mean.

**Comment by GitMensch** [comment]
Would there be any "thickness" it the objdump colored output that may should be copied over?

Note: in any case I'd "expect" to have the objdump coloring as an extra entry of the highlighting (create a manual list including all syntax highlighters that have "asm" in them, the add "objdump" on top, if it is available).
Last time checked there the original syntax highlighter doesn't work any more for disassembly at all.

**Comment by GitMensch** [comment]
If it is more likely to get included faster, then we may split this PR in 4:
* separate hex column - at least according to my manual tests this seems to work fine
* a later follow-up PR for a possible automated testcase for the above
* move to QTextLayout including caching (seems to be finished and provide better performance - in this case an early distribution would be very useful) -> possibly in this PR
* disassembly syntax highlighting with objdump

**Comment by milianw** [comment]
code lgtm, probably needs a rebase now but otherwise this should be ready to go in IIUC? @GitMensch is your highlighting bug resolved now in the latest iteration of this pull request?

**Comment by GitMensch** [comment]
> GitMensch is your highlighting bug resolved now in the latest iteration of this pull request?

No, sadly #561 is not yet solved. Just tested with [objdump.log](https://github.com/KDAB/hotspot/files/13629286/objdump.cob_move_packed_to_display.log), generated by "GNU objdump (GNU Binutils for Debian) 2.35.2", which doesn't support `--disassembler-color`, with `objdump -d -l -C --start-address 0x2ba15 --stop-address 0x2bccf --visualize-jumps /tmp/gnucobol-bug-hunting/libcob/.libs/libcob.so.4.2.0`, result 
broken disassembly

I guess that the actual highlighting is not the main problem but the splitting of the jump-visualisation and hex-values.

I'm keen to recheck after a rebase (shouldn't make a big difference, I guess) and even more after a fix to this.

Just a debugging note (you're likely aware of this): you may add an "objdump" script that only does a `cat` on this file, to see the result in your environment.

**Comment by lievenhey** [comment]
Thanks, I finally understand your problem and managed to reproduce it.

**Comment by lievenhey** [comment]
@GitMensch It seems that #566 fixed the issue, can you check again?

**Comment by GitMensch** [comment]
Yes, it is and the current appimage from this PR works fine there, too.
Also tested with a bunch of other recordings. 🥳 

As Milian is also fine with this PR it seems its ready for merge :-)

**Comment by GitMensch** [comment]
@milianw can you merge that in, please?

**Review by milianw — APPROVED** [review]
